File: sdks/swift/Sources/Modula/IDs.swift
import Foundation

// MARK: - ResourceID Protocol

public protocol ResourceID: Codable, Sendable, Hashable, CustomStringConvertible, ExpressibleByStringLiteral {
    var rawValue: String { get }
    init(_ rawValue: String)
}

public extension ResourceID {
    var description: String { rawValue }
    var isZero: Bool { rawValue.isEmpty }

    init(stringLiteral value: String) {
        self.init(value)
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        let value = try container.decode(String.self)
        self.init(value)
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        try container.encode(rawValue)
    }
}

// MARK: - Content IDs

public struct ContentID: ResourceID {
    public let rawValue: String
    public init(_ rawValue: String) { self.rawValue = rawValue }
}

public struct ContentFieldID: ResourceID {
    public let rawValue: String
    public init(_ rawValue: String) { self.rawValue = rawValue }
}

public struct ContentRelationID: ResourceID {
    public let rawValue: String
    public init(_ rawValue: String) { self.rawValue = rawValue }
}

// MARK: - Admin Content IDs

public struct AdminContentID: ResourceID {
    public let rawValue: String
    public init(_ rawValue: String) { self.rawValue = rawValue }
}

public struct AdminContentFieldID: ResourceID {
    public let rawValue: String
    public init(_ rawValue: String) { self.rawValue = rawValue }
}

public struct AdminContentRelationID: ResourceID {
    public let rawValue: String
    public init(_ rawValue: String) { self.rawValue = rawValue }
}

// MARK: - Schema IDs

public struct DatatypeID: ResourceID {
    public let rawValue: String
    public init(_ rawValue: String) { self.rawValue = rawValue }
}

public struct DatatypeFieldID: ResourceID {
    public let rawValue: String
    public init(_ rawValue: String) { self.rawValue = rawValue }
}

public struct FieldID: ResourceID {
    public let rawValue: String
    public init(_ rawValue: String) { self.rawValue = rawValue }
}

// MARK: - Admin Schema IDs

public struct AdminDatatypeID: ResourceID {
    public let rawValue: String
    public init(_ rawValue: String) { self.rawValue = rawValue }
}

public struct AdminDatatypeFieldID: ResourceID {
    public let rawValue: String
    public init(_ rawValue: String) { self.rawValue = rawValue }
}

public struct AdminFieldID: ResourceID {
    public let rawValue: String
    public init(_ rawValue: String) { self.rawValue = rawValue }
}

// MARK: - Field Type IDs

public struct FieldTypeID: ResourceID {
    public let rawValue: String
    public init(_ rawValue: String) { self.rawValue = rawValue }
}

public struct AdminFieldTypeID: ResourceID {
    public let rawValue: String
    public init(_ rawValue: String) { self.rawValue = rawValue }
}

// MARK: - Media IDs

public struct MediaID: ResourceID {
    public let rawValue: String
    public init(_ rawValue: String) { self.rawValue = rawValue }
}

public struct MediaDimensionID: ResourceID {
    public let rawValue: String
    public init(_ rawValue: String) { self.rawValue = rawValue }
}

// MARK: - Auth IDs

public struct UserID: ResourceID {
    public let rawValue: String
    public init(_ rawValue: String) { self.rawValue = rawValue }
}

public struct RoleID: ResourceID {
    public let rawValue: String
    public init(_ rawValue: String) { self.rawValue = rawValue }
}

public struct SessionID: ResourceID {
    public let rawValue: String
    public init(_ rawValue: String) { self.rawValue = rawValue }
}

public struct TokenID: ResourceID {
    public let rawValue: String
    public init(_ rawValue: String) { self.rawValue = rawValue }
}

public struct UserOauthID: ResourceID {
    public let rawValue: String
    public init(_ rawValue: String) { self.rawValue = rawValue }
}

public struct UserSshKeyID: ResourceID {
    public let rawValue: String
    public init(_ rawValue: String) { self.rawValue = rawValue }
}

public struct PermissionID: ResourceID {
    public let rawValue: String
    public init(_ rawValue: String) { self.rawValue = rawValue }
}

public struct RolePermissionID: ResourceID {
    public let rawValue: String
    public init(_ rawValue: String) { self.rawValue = rawValue }
}

// MARK: - Routing IDs

public struct RouteID: ResourceID {
    public let rawValue: String
    public init(_ rawValue: String) { self.rawValue = rawValue }
}

public struct AdminRouteID: ResourceID {
    public let rawValue: String
    public init(_ rawValue: String) { self.rawValue = rawValue }
}

// MARK: - Other IDs

public struct TableID: ResourceID {
    public let rawValue: String
    public init(_ rawValue: String) { self.rawValue = rawValue }
}

public struct EventID: ResourceID {
    public let rawValue: String
    public init(_ rawValue: String) { self.rawValue = rawValue }
}

public struct BackupID: ResourceID {
    public let rawValue: String
    public init(_ rawValue: String) { self.rawValue = rawValue }
}

// MARK: - Value Types

public struct Slug: ResourceID {
    public let rawValue: String
    public init(_ rawValue: String) { self.rawValue = rawValue }
}

public struct Email: ResourceID {
    public let rawValue: String
    public init(_ rawValue: String) { self.rawValue = rawValue }
}

public struct URLValue: ResourceID {
    public let rawValue: String
    public init(_ rawValue: String) { self.rawValue = rawValue }
}

// MARK: - Timestamp

public struct Timestamp: ResourceID {
    public let rawValue: String
    public init(_ rawValue: String) { self.rawValue = rawValue }

    private static let formatter: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime]
        return f
    }()

    public func date() -> Date? {
        Timestamp.formatter.date(from: rawValue)
    }

    public init(date: Date) {
        self.rawValue = Timestamp.formatter.string(from: date)
    }

    public static func now() -> Timestamp {
        Timestamp(date: Date())
    }
}

File: sdks/typescript/types/src/ids.ts
/**
 * Branded ID types for compile-time safety across all ModulaCMS entities.
 *
 * @module ids
 */

import type { Brand } from './brand.js'

/** Unique identifier for a user account. */
export type UserID = Brand<string, 'UserID'>

/** Unique identifier for an admin content data node. */
export type AdminContentID = Brand<string, 'AdminContentID'>

/** Unique identifier for an admin content field value. */
export type AdminContentFieldID = Brand<string, 'AdminContentFieldID'>

/** Unique identifier for an admin content relation. */
export type AdminContentRelationID = Brand<string, 'AdminContentRelationID'>

/** Unique identifier for an admin datatype definition. */
export type AdminDatatypeID = Brand<string, 'AdminDatatypeID'>

/** Unique identifier for an admin field definition. */
export type AdminFieldID = Brand<string, 'AdminFieldID'>

/** Unique identifier for an admin route. */
export type AdminRouteID = Brand<string, 'AdminRouteID'>

/** Unique identifier for a public content data node. */
export type ContentID = Brand<string, 'ContentID'>

/** Unique identifier for a public content field value. */
export type ContentFieldID = Brand<string, 'ContentFieldID'>

/** Unique identifier for a public content relation. */
export type ContentRelationID = Brand<string, 'ContentRelationID'>

/** Unique identifier for a datatype (schema) definition. */
export type DatatypeID = Brand<string, 'DatatypeID'>

/** Unique identifier for a field (schema) definition. */
export type FieldID = Brand<string, 'FieldID'>

/** Unique identifier for a media asset. */
export type MediaID = Brand<string, 'MediaID'>

/** Unique identifier for a user role. */
export type RoleID = Brand<string, 'RoleID'>

/** Unique identifier for a permission. */
export type PermissionID = Brand<string, 'PermissionID'>

/** Unique identifier for a role-permission junction row. */
export type RolePermissionID = Brand<string, 'RolePermissionID'>

/** Unique identifier for a field type lookup entry. */
export type FieldTypeID = Brand<string, 'FieldTypeID'>

/** Unique identifier for an admin field type lookup entry. */
export type AdminFieldTypeID = Brand<string, 'AdminFieldTypeID'>

/** Unique identifier for a public route. */
export type RouteID = Brand<string, 'RouteID'>

/** Unique identifier for an active session. */
export type SessionID = Brand<string, 'SessionID'>

/** Unique identifier for a user OAuth connection. */
export type UserOauthID = Brand<string, 'UserOauthID'>

/** URL-safe slug string used to identify routes. */
export type Slug = Brand<string, 'Slug'>

/** Branded email address string. */
export type Email = Brand<string, 'Email'>

/** Branded URL string. */
export type URL = Brand<string, 'URL'>
